File: Algorithm_5_Longest Palindromic Substring.cpp
#include <iostream>
using namespace std;
class Solution {
public:
    int low=0, maxLen=1;
    string longestPalindrome(string s) {
        if (s.size()<2) return s;
        for (int i=0;i<s.size()-1;i++) getPalindrome(s,i);
        return s.substr(low,maxLen);
    }
    void getPalindrome(string s, int i)
    {
        int l,h;
        for (int k=0;k<2;k++)
        {
            if (k==0) {l=i;h=i;}
            else {l=i;h=i+1;}
             while (l>=0 && h<s.size() && s[l]==s[h]) {l--;h++;}
            l++;
            h--;
            if (h-l+1>maxLen)
            {
                low=l;
                maxLen=h-l+1;
            }
        }
    }
};

File: Algorithm_5_Longest Palindromic Substring.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        if not s:
            return ""
        dp = [[0 for i in range(len(s))] for i in range(len(s))]
        maxLen = maxStart = maxEnd = 0
        for length in range(len(s)):
            for curStart in range(len(s) - length):
                if length <= 2:
                    dp[curStart][curStart + length] = 1 if s[curStart] == s[curStart + length] else 0
                else:
                    dp[curStart][curStart + length] = 1 if s[curStart] == s[curStart + length] and dp[curStart + 1][curStart + length - 1] else 0
                if dp[curStart][curStart + length] and length + 1 > maxLen:
                    maxLen = length + 1
                    maxStart = curStart
                    maxEnd = curStart + length
        return s[maxStart:maxEnd + 1]

    def longestPalindrome1(self, s:str) -> str:
        if not s:
            return ""
        res = ""
        for i in range(len(s)):
            tmp = self.helper(s, i, i)
            if len(tmp) > len(res):
                res = tmp
            tmp = self.helper(s, i, i + 1)
            if len(tmp) > len(res):
                res = tmp
        return res

    def helper(self, s:str, left:int, right:int) -> str:
        while left >= 0 and right < len(s) and s[left] == s[right]:
            left -= 1
            right += 1
        return s[left + 1:right]